### DISCUSSION TITLE
How to unhide ocr layer created by Adobe?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#3013] [General]
[2.pdf](https://github.com/pymupdf/PyMuPDF/files/13899588/2.pdf)

![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/193c29e6-bbb5-47ed-8ec1-426d31ae9be9)

Attached a file with Adobe OCR layer.
I want to get below info.
Textline box, font info, font name, alignment etc. so that I can imitate when adding my own result.

**Comment by JorjMcKie** [comment]
First of all, thanks for your understanding!

You can do something this:
```python
import fitz
doc=fitz.open("2.pdf")
page=doc[0]
blocks=page.get_text("dict",flags=fitz.TEXTFLAGS_TEXT)["blocks"]
spans=[]
for b in blocks:
    for l in b["lines"]:
        spans.extend([s for s in l["spans"]])

        
spans[0]
{'size': 10.199999809265137, 'flags': 4, 'font': 'Times-Roman', 'color': 0, 'ascender': 1.0529999732971191, 'descender': -0.2809999883174896, 'text': 'ICS 35. 040 ', 'origin': (54.16999816894531, 53.26995849609375), 'bbox': (54.16999816894531, 42.52935791015625, 109.11921691894531, 56.13615798950195)}
spans[-1]
{'size': 28.48898696899414, 'flags': 4, 'font': 'HiddenHorzOCR', 'color': 0, 'ascender': 1.0, 'descender': -0.20000000298023224, 'text': '中华人民共和国水利部发布', 'origin': (140.1699981689453, 777.5999755859375), 'bbox': (140.1699981689453, 746.8800048828125, 457.2100830078125, 783.7439575195312)}
```

Text **_alignment_** is no information stored in any PDF. All you can have is a variant of the above that delivers text by character ("rawdict" instead of "dict"). Then use the delivered single character bboxes to rewrite the text character-by-character.

**Comment by nissansz** [comment]
For TextWriter morph,
the only way to use morph is to calculate text_length/rect_width for autofit?

How to correctly calculate text_length in pixels?
My calculation result seems wrong, the written textline does not match.
 font is windows deng.ttf

fontsize = 15


            textwidth = fitz.get_text_length(text, fontname = 'china-s',  fontsize=fontsize, encoding='utf-8')
            # pivot = fitz.Point(rect.x0, rect.y0 + rect.height / 2)  # middle of left side
            # pos = pivot
            mat = fitz.Matrix(textwidth/rect.width, 1.0)
            # mat = fitz.Matrix(1.0, 1.0)
            # mat = fitz.Matrix(rect.width/textwidth, 1.0)

            # mat = fitz.Matrix(1/0.7, 1)  # 2 = double the width
            tw.write_text(page, opacity=None, color=color, morph=(pos, mat), overlay=True, oc=0, render_mode=0)
            
[3_overlay_tw.pdf](https://github.com/pymupdf/PyMuPDF/files/13900735/3_overlay_tw.pdf)

result is as below
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/837140d1-459d-45e5-b485-322f7664af59)

**Reply by JorjMcKie** [reply]
Multiple comments:

* TextWriter **never** uses Base-14 fonts, it always embeds font files (see documentation!).
* for Courier, Helvetica, Times-Roman it uses font files with equal metrix as the Base-14 equivalents, but not for CJK. So you cannot correctly measure text length as you do.
* In your use case, you seem to be interested in one-line outputs only. Therefore, it makes not sense to use fill_textbox method. Instead, make a separate TextWriter for each line and use `TextWriter.append()`. There is a TextWriter property `last_point` (see documentation), which gives you the end position after each `append`. Then compute the difference to the start and you have the effective text length - whatever font may have been used.

**Comment by nissansz** [comment]
Thank you.
I found a solution.
It seesms working well.

https://stackoverflow.com/questions/75189505/pymupdf-get-optimal-font-size-given-a-rectangle
fs = rect.width / font.text_length(text, fontsize=1)

**Comment by nissansz** [comment]
Is it possible to reduce font size to half, and stretch fit with morph?
'''
                    tw = fitz.TextWriter(page.rect)
                    origin = (int(updateleft), int(top-5))
                    pos = fitz.Point(updateleft, top-5)
                    mat = fitz.Matrix(fontsize_fit/fontsize, 1.0)
                    tw.append(origin, clean_char, fontsize=fontsize, font=font, small_caps=True)
'''
![image](https://github.com/user-attachments/assets/c5721bbc-70d9-4516-805d-6ca0ea07057c)

**Comment by nissansz** [comment]
Another problem to use character writer is there are additional spaces between each english letter.
v i d e o d e c o d i n g
